### DISCUSSION TITLE
lit playground: navigator.serviceWorker is undefined in firefox private mode

### DISCUSSION THREAD

**Discussion by mwmcode** [lit/lit#2217] [General]
Hello,

Using Firefox (private mode), I was checking out https://lit.dev/playground/ and they playground was in infinite loading state 

```js
// console output
Uncaught (in promise) TypeError: navigator.serviceWorker is undefined
```
Then I checked [MDN's Navigator.serviceWorker](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/serviceWorker), and it states:

"The Navigator.serviceWorker read-only property returns the ServiceWorker**Container** object ...The feature may not be available in **private mode**."

I'm curious to why using the Container here? any benefits vs. using navigator.servieWorker? workaround for such scenario (user in private mode)?

Thanks,

**Comment by aomarks** [comment]
Hi @mcha-dev. You are correct that Firefox does not support Service Workers in private mode (unlike Chrome and Safari, which do). See https://bugzilla.mozilla.org/show_bug.cgi?id=1320796 for the bug on Firefox, and https://github.com/google/playground-elements/issues/78 to track adding a nicer warning to the Playground for when this happens.

The reason we use Service Workers is because it is key to the implementation of https://github.com/google/playground-elements, which is the library that powers the lit.dev playground. Service Workers allow us to host a virtual file server that serves the files as you edit them, hosted entirely within the browser, so that there is no traditional backend server needed. This is great for performance (since there is no network overhead) and privacy (since nothing you type goes to a remote server).

**Reply by mwmcode** [reply]
I think MDN could convey that it's a bug in the docs, the current phrasing makes it seem like _a behavior to be expected_


> allow us to host a virtual file server that serves the files as you edit them, hosted entirely within the browser, so that there is no traditional backend server needed. This is great for performance (since there is no network overhead) and privacy (since nothing you type goes to a remote server).

💡TIL. That's pretty cool! 

Thanks @aomarks

**Reply by aomarks** [reply]
Update: playground-elements and lit.dev will now show a more helpful error about why the preview doesn't work in Firefox private browsing mode:

<img src="https://user-images.githubusercontent.com/48894/136377927-8556ceab-6573-4b1d-ac9e-a5a288118b2b.png" width="600px">
